client/session.go

```diff
@@ -574,99 +651,67 @@ func (s *session) Write(namespace, id string, t time.Time, value float64, unit x
-		var snapshotSuccess int32
-		if err != nil {
-			atomic.AddInt32(&resultErrs, 1)
-			resultErrLock.Lock()
-			errors = append(errors, err)
-			resultErrLock.Unlock()
-		} else {
-			snapshotSuccess = atomic.AddInt32(&success, 1)
-		}
-		remaining := atomic.AddInt32(&pending, -1)
-		doneAll := remaining == 0
-		switch s.writeLevel {
-		case topology.ConsistencyLevelOne:
-			complete := snapshotSuccess > 0 || doneAll
-			if complete && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
-				wg.Done()
-			}
-		case topology.ConsistencyLevelMajority:
-			complete := snapshotSuccess >= majority || doneAll
-			if complete && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
-				wg.Done()
-			}
-		case topology.ConsistencyLevelAll:
-			if doneAll && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
-				wg.Done()
-			}
-		}
-		// Finalize resources that can be returned
-		if doneAll {
-			// Return write to pool
-			s.writeOpPool.Put(w)
-		}
-	}
+	c.op, c.majority = s.writeOpPool.Get(), majority
 
-	routeErr := s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
+	c.op.namespace = ts.StringID(namespace)
+	c.op.request.ID = tsID.Data()
+	c.op.request.Datapoint.Value = value
+	c.op.request.Datapoint.Timestamp = timestamp
+	c.op.request.Datapoint.TimestampType = timeType
+	c.op.request.Datapoint.Annotation = annotation
+	c.op.completionFn = c.completionFn
+
+	if err := s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
 		// First count all the pending write requests to ensure
 		// we count the amount we're going to be waiting for
 		// before we enqueue the completion fns that rely on having
 		// an accurate number of the pending requests when they execute
-		pending++
-	})
-	if routeErr == nil {
-		// Now enqueue the write requests
-		routeErr = s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
-			enqueued++
-			if err := s.queues[idx].Enqueue(w); err != nil && enqueueErr == nil {
-				// NB(r): if this ever happens we have a code bug, once we
-				// are in the read lock the current queues we are using should
-				// never be closed
-				enqueueErr = err
-			}
-		})
+		c.pending++
+		c.queues = append(c.queues, s.queues[idx])
+	}); err != nil {
+		c.decref()
+		s.RUnlock()
+		return err
 	}
-	s.RUnlock()
 
-	if routeErr != nil {
-		return routeErr
-	}
+	c.Lock()
 
-	if enqueueErr != nil {
-		s.opts.InstrumentOptions().Logger().Errorf("failed to enqueue write: %v", enqueueErr)
-		return enqueueErr
+	for i := range c.queues {
+		if err := c.queues[i].Enqueue(c.op); err != nil {
+			c.Unlock()
+			c.decref()
+
+			// NB(r): if this ever happens we have a code bug, once we
+			// are in the read lock the current queues we are using should
+			// never be closed
+			s.RUnlock()
+			s.opts.InstrumentOptions().Logger().Errorf("failed to enqueue write: %v", err)
+			return err
+		}
+
+		c.incref()
+		enqueued++
 	}
 
 	// Wait for writes to complete
-	wg.Wait()
+	s.RUnlock()
+	c.condition.Wait()
```

Comment:
Spurious wake ups don't happen because Golang's condvars has nothing to do with pthreads' condvars. According to Golang authors, `It's built on Semacquire, which does not permit spurious interrupts`.